### ISSUE TITLE
Allow performing verification upon first resolve

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#555]
Many inexperienced users forget to call `Verify()` after configuring the container. To make it easier to do the right thing, we should consider automatically calling `Verify()` when the first service is resolved, in case the user didn't call it. 

Since verification can cause performance problems for big applications, auto-verification should be configurable, for instance:

``` C#
var container = new Container();
container.Options.EnableAutoVerification = true; // default false in v4.x

container.Register<IFoo, FooImpl>();

container.GetInstance<IFoo>(); // will trigger verification when EnableAutoVerification = true
```

Having the feature enabled by default, however, is a major breaking change, because in case the user upgrades to this new version, without setting `EnableAutoVerification` to `false`, it will cause the container to be verified, even though the user might explicitly prevented this in his production environment:

``` c#
#if DEBUG
    // Prevent verification in release mode.
    container.Verify();
#endif
```

In v4 the feature can be added, as long as it is off by default.

**Comment by TheBigRic** [comment]
I'm tempted to react with thumbs up because verification is something which separates Simple Injector from the rest and is also some kind of ultra climax of all [design principles](https://simpleinjector.readthedocs.io/en/latest/principles.html).

However. I think the impact of the breaking change is huge and maybe too huge!

Maybe there is an elegant solution which could be used in two ways. What do you think of including a .ps script in the NuGet package which generates some commented code? This could be both 

``` c#
container.Verify();
```
and
```c#
container.Options.EnableAutoVerification = true;
```
of course shipped as a more or less complete bootstrapper.cs/bootstrapper.vb

**Comment by Tornhoof** [comment]
@TheBigRic <PackageReference> nuget  references can't use install.ps1 anymore. So this only works for the legacy references.

**Comment by TheBigRic** [comment]
@Tornhoof 
Ok, good to know this is not supported anymore. However, what install.ps1 was meant to do as a feature is still there although somewhat different. 

At this time it is probably more interesting to decide which path to take with this, to auto verify or not to auto verify, that's the question :-)

**Comment by dotnetjunkie** [comment]
[feature-555](https://github.com/simpleinjector/SimpleInjector/tree/feature-555) branch created.

**Comment by dotnetjunkie** [comment]
Hi @davidroth, you voted with a thumbs down. Can you elaborate on why you think this is feature is a bad idea?

**Comment by davidroth** [comment]
@dotnetjunkie Not sure, too long ago :-) After re-reading I think it's quite a good idea actually ;-)

Did the issue description always include the EnableAutoVerification setting? :-) 

Removed my downvote - thanks for pinging me ;-)

**Comment by dotnetjunkie** [comment]
@davidroth I was trying to find out what the counter arguments would be for implementing this feature, as I now started working on implementing it. I suspect you downvoted because this feature was initially described as a breaking change (which it still is, but the breaking part is moved to #747). The issue has been updated several times, so it's very well possible that it was confusing to begin with.
